lib/devise-security/hooks/session_traceable.rb

```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+# After each sign in, create new traceable session.
+# This is only triggered when the user is explicitly set (with set_user)
+# and on authentication. Retrieving the user from session (:fetch) does
+# not trigger it.
+Warden::Manager.after_set_user except: :fetch do |record, warden, options|
+  scope = options[:scope]
+  if record.devise_modules.include?(:session_traceable) &&
+     warden.authenticated?(options[:scope]) &&
+     !record.skip_session_traceable?
+    opts = {
+      ip_address: warden.request.remote_ip,
+      user_agent: warden.request.headers['User-Agent']
+    }
+    unique_traceable_token = record.log_traceable_session!(opts)
+    if unique_traceable_token.present?
+      warden.session(options[:scope])['unique_traceable_token'] = unique_traceable_token
+    else
+      warden.logout(scope)
+      throw(:warden, scope: scope, message: :unauthenticated)
+    end
+  end
+end
+
+# Each time a record is fetched from session we verify if the session
+# has a valid unique session identifier. If so, then we set the last accessed time.
+Warden::Manager.after_set_user only: :fetch do |record, warden, options|
+  scope = options[:scope]
+  env = warden.request.env
+  opts = { ip_address: warden.request.remote_ip }
+  if record.devise_modules.include?(:session_traceable) &&
+     warden.authenticated?(scope) &&
+     options[:store] != false &&
```

Comment:
By default, `store` is always a `NilClass` unless `warden.authenticate` is called with an option `store: false` or on `devise` calling `authenticate_user` with `store: false` so it will be safer to use that condition which follows the [convention](https://github.com/heartcombo/devise/blob/1d6658097e364d45b5e059976f1e822eee7d67da/lib/devise/hooks/rememberable.rb#L5) in `devise`.